fgprof.go

```diff
@@ -75,6 +75,11 @@ func (p *profiler) GoroutineProfile() []runtime.StackRecord {
 	// TODO(fg) There might be workloads where it would be nice to shrink
 	// p.stacks dynamically as well, but let's not over-engineer this until we
 	// understand those cases better.
+
+	nGoroutines := runtime.NumGoroutine()
+	if len(p.stacks) < nGoroutines {
+		p.stacks = make([]runtime.StackRecord, int(float64(nGoroutines)*1.1))
+	}
```

Comment:
Hm, I was thinking about doing this outside the `GoroutineProfile()` func, otherwise it's costing us extra cycles even after we've settled on a suitable large `stacks` size. (I wasn't able to see a significant effect from this in the benchmark, but I think it's theoretically clear why this is higher overhead)

Maybe something like this? And then changing all the places where we initialize a &profiler{} struct right now?

```go
// newProfiler returns a new profiler with a pre-allocated stacks field. This
// is slightly faster than discovering its size during the first
// GoroutineProfile() call.
func newProfiler() *profiler {
	n := runtime.NumGoroutine()
	return &profiler{
		stacks: make([]runtime.StackRecord, int(float64(n)*1.1)),
	}
}
```

Let me know what you think!

Reply:
Awesome! Do you have interest/time in modifying your PR as suggested?

Reply:
Well, I have done few tests on this.

The issue is: `runtime.NumGoroutine` is faster than calling `GoroutineProfile` and it optimizes the use case where there are always more goroutines than `p.stacks`. However, as you indicated this makes it bad for the situation where `goroutines` count is similar, I did not know `p.stacks` is reused when I opened the case.

Here is a sample implementation for the above one, but IMHO: it does not make any sense. It adds more complexity with too little gain.

I propose to close this issue as is and focus on maybe more important things in the hot path, or if you like I can tidy the PR up and we can merge.

Reply:
> Well, I have done some further tests on this.
> 
> The issue is: as `runtime.NumGoroutine` is faster than calling `GoroutineProfile` I thought it would make sense to optimize for this as I opened the case.But I did not know `p.stacks` is reused. My previous implementation was still not good enough as it still adds more cycles where goroutine counts do not change between calls to `GoroutineProfile`.
> 
> I pushed a sample implementation for the above one, but it seems like it does not make too much difference because the `newProfile` is not on the hot path anymore only initialized once. It adds more complexity with too little gain.

Yeah, I agree that it won't have a big impact. So if you feel like this is a bad tradeoff I'd also prefer to not merge the patch

> So, it is up to you:
> 
> * We can close this issue as is and focus on maybe more important things in the hot path,
> * We can merge,

Ok, I think we should close this and look for other opportunities.

I think the main issue with fgprof right now is that it doesn't scale to programs that use tons of goroutines. I'm starting a new job at Datadog's Profiler team in January that might give me the opportunity to submit proposals and patches to the Go core for it.

But meanwhile it might be nice for the profiler to detect if it's too slow (and causing issues for the profiled program). When that happens we could either dynamically adjust the sampling rate, or fail with an error.

Let me know if you're interested in playing with this idea.

Also, if you change your mind and want to merge the PR, I'd be okay with that too. It's not a lot of complexity to add IMO.

Reply:
> Let me know if you're interested in playing with this idea.

Absolutely yes! Would love to see what kind of improvements can be done. One of the tricky parts is also: how to detect profiler is too slow itself, do you have any idea on that? From my past experience, for example I know that if you do everything right(meaning use good data structures O(1), have good cache locality..etc.etc), then somehow the bootleneck becomes reading the underlying CPU clock which is not a light operation at all.

So, I am all in for the subject and would love to discuss ideas with you...

Reply:
> > Let me know if you're interested in playing with this idea.
> 
> Absolutely yes! Would love to see what kind of improvements can be done.

🥳

> One of the tricky parts is: how to detect profiler is too slow itself, do you have any idea on that? From my past experience, for example I know that if you do everything right(meaning use good data structures O(1), have good cache locality..etc.etc), then somehow the bootleneck becomes reading the underlying CPU clock which is not a light operation at all.

So my naive idea is something like this: We want to guarantee that fgprof uses X% of the programs time at most. Let's say 1%. This means that we get 10ms "fgprof time" per 1s of program execution. So we could have a simple `time.Duration` variable that sums up how much `GoroutineProfile()` time we've been using and check its value every 1s to make sure it doesn't exceed `10ms`. Based on this we could either stop profiling, or adapt the frequency.

I don't think we'll run into issues with CPU clock overhead, as we'll check the clock 198 (2*99) times per second at most. That shouldn't be an issue.

Of course there is plenty of room to get fancy. E.g. using exponentially weighted averages for tracking the overhead or PID controllers for adapting the frequency.

But perhaps trying the most stupid/simple approach outlined above and stopping the profiling when its too slow could be a good starting point for experimentation.

Let me know what you think.

Reply:
Hmmm... Now, I need/will be thinking for this inside my head for a while! You might be right about CPU clock issues as my experience mostly comes from tracing profilers, statistical profilers is different. 

I will follow up.

Reply:
One thing I could not understand:

> So we could have a simple time.Duration variable that sums up how much GoroutineProfile() time we've been using and check its value every 1s to make sure it doesn't exceed 10ms

How we calculate how much time an application is spending in total? Like suppose we have millions of goroutines waiting  for I/O doing not much. Will we aim for %X in that case or 10ms?

Reply:
My mental model is this: Calling `runtime.GoroutineProfile()` calls `stopTheWorld()` so any time we spent inside of it, we assume the users application is totally stopped, and we're "stealing" its time for profiling. The goal is to not steal more than e.g 1% of the programs wall clock time.

We don't really know how stuff the program is doing when not profiling, but I don't think it's relevant for the goal of the profiler to minimize interfering with the program.

Reply:
Now I understand. I forgot about the STW in `goroutineprofile`. I am still in learning phase for the runtime. Now it makes  sense also looking at the wall time as well (instead of CPU time).

I will do some experiments myself and think a bit more...

Reply:
I was too busy lately and could not find time to look/think into this too much.

But I thought about the problem a bit more and I agree that your approach of trying to bound profiler wall time into %1 seems both simple to implement and powerful enough like you suggested.

Couple of issues that came to my mind:

1/ While I have not tested it yet: I assume when STW occurs, it might have some adverse effects on the system as it will abuse cache locality of some processors. One test that we could either do manually or automatically is to have a large number of CPU intensive goroutines executing and do some work with them. THen we calculate these both with profiler and without profiler and see if the output is also changed like %1 or so. It would be nice to test this as there might be other kind of adverse effects as well.

2/ I think this is more major problem: what happens if `GoroutineProfile` itself is actually taking more than %1? Yeah, we could simply skip some samples on the next iterations but that might actually hurt accuracy? What if it took 30ms? 50ms? 100ms? Those numbers are just (maybe?) too big for the application? Well again: I have not yet tested it myself but I think
this is not too uncommon if you have many goroutines: https://github.com/golang/go/issues/33250 Well one idea might be to make an estimation on how long a `GoroutineProfile` might take based on previous data, but again: I don't know what we will do even if that is the case. One idea: it might a good idea to have some help from the runtime: like what if `GoroutineProfile` returns some data even if `len(p)<=n`. Even we don't have all the goroutine data, at least we might have something(maybe some kind of randomness can be applied in the runtime for returning different GOROs everytime?). So, instead of returning nothing, we will have at least something to profile this way. Otherwise it is completely binary, right?
Any idea on this?

Reply:
@sumerc this PR is an odd place for this discussion, so I've created a new issue and will reply there: https://github.com/felixge/fgprof/issues/12

If you're okay with it, I think we should close this PR for now?

Reply:
Sure.